File: lib/excel_generator.py
import os
import json
import sys
import pandas as pd
from datetime import datetime

# Define the required fields
REQUIRED_FIELDS = [
    'ACTIVITYID',
    'DATECREATED',
    'DATEUPDATED',
    'DATEUSERFIELDUPDATED',
    'INTERVIEWSCHEDULEDATE',
    'CANDIDATEID',
    'CANDIDATEFIRSTNAME',
    'CANDIDATELASTNAME',
    'CANDIDATEEMAIL',
    'JOBREFERENCENUMBER',
    'JOBTITLE',
    'DIVISION',
    'COMPANYNAME',
    'SUBMITTALDATE',
    'INTERVIEWDATE',
    'INTERVIEW_TIMEZONEID',
    'L1 Interview Status'
]

def generate_excel(data):
    """Generate Excel file from JobDiva data"""
    try:
        # Print data structure for debugging
        print(json.dumps({'debug': 'Data structure received:', 'data': data[:1] if isinstance(data, list) else data}, indent=2))
        
        # Convert data to DataFrame
        if isinstance(data, dict) and 'data' in data:
            # If data is wrapped in a response object
            df = pd.DataFrame(data['data'])
        else:
            # If data is directly an array
            df = pd.DataFrame(data)
        
        # Print columns for debugging
        print(json.dumps({'debug': 'Available columns:', 'columns': df.columns.tolist()}))
        
        # Check if the required column exists
        if 'L1 Interview Status' not in df.columns:
            raise ValueError(f"Column 'L1 Interview Status' not found. Available columns: {df.columns.tolist()}")
        
        # Filter for Selected for L2 status
        filtered_df = df[df['L1 Interview Status'] == 'Selected for L2']
        
        # Select only the required fields that exist in the DataFrame
        existing_fields = [field for field in REQUIRED_FIELDS if field in filtered_df.columns]
        filtered_df = filtered_df[existing_fields]
        
        # Create Excel_Tables directory if it doesn't exist
        os.makedirs('Excel_Tables', exist_ok=True)
        
        # Use fixed filename
        filename = "DemoDatabase2.xlsx"
        output_path = os.path.join('Excel_Tables', filename)
        
        # Save to Excel
        filtered_df.to_excel(output_path, index=False)
        
        result = {
            'success': True,
            'filename': filename,
            'path': output_path,
            'record_count': len(filtered_df),
            'total_records': len(df),
            'filtered_records': len(filtered_df),
            'fields_included': existing_fields,
            'missing_fields': [field for field in REQUIRED_FIELDS if field not in existing_fields]
        }
        print(json.dumps({'debug': 'Operation completed successfully', 'result': result}, indent=2))
        return result
    except Exception as e:
        error_msg = f"Error generating Excel: {str(e)}"
        print(json.dumps({'error': error_msg}, indent=2))
        return {
            'success': False,
            'error': error_msg
        }

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print(json.dumps({'success': False, 'error': 'Input file path is required'}))
        sys.exit(1)
        
    input_file = sys.argv[1]
    
    try:
        print(json.dumps({'debug': f'Reading input file: {input_file}'}, indent=2))
        with open(input_file, 'r') as f:
            data = json.load(f)
        
        result = generate_excel(data)
        print(json.dumps(result))
        sys.exit(0 if result['success'] else 1)
    except Exception as e:
        error_msg = f"Script error: {str(e)}"
        print(json.dumps({'success': False, 'error': error_msg}, indent=2))
        sys.exit(1) 

File: lib/excelGenerator.ts
import * as XLSX from 'xlsx';
import path from 'path';
import fs from 'fs';
import os from 'os';

interface ExcelResult {
  success: boolean;
  filename?: string;
  buffer?: Buffer;
  record_count?: number;
  total_records?: number;
  filtered_records?: number;
  fields_included?: string[];
  missing_fields?: string[];
  error?: string;
}

// Define the required fields
const REQUIRED_FIELDS = [
  'ACTIVITYID',
  'DATECREATED',
  'DATEUPDATED',
  'DATEUSERFIELDUPDATED',
  'INTERVIEWSCHEDULEDATE',
  'CANDIDATEID',
  'CANDIDATEFIRSTNAME',
  'CANDIDATELASTNAME',
  'CANDIDATEEMAIL',
  'JOBREFERENCENUMBER',
  'JOBTITLE',
  'DIVISION',
  'COMPANYNAME',
  'SUBMITTALDATE',
  'INTERVIEWDATE',
  'INTERVIEW_TIMEZONEID',
  'L1 Interview Status'
];

export async function generateExcel(data: any): Promise<ExcelResult> {
  try {
    console.log('Starting Excel generation');
    
    // Convert data to array if it's wrapped in a response object
    const records = Array.isArray(data) ? data : (data.data || []);
    
    // Calculate the date 5 days prior to the toDate
    const toDate = new Date(); // Assuming toDate is today
    const fiveDaysPrior = new Date(toDate);
    fiveDaysPrior.setDate(toDate.getDate() - 5);

    // Filter for Selected for L2 status, Capgemini India, and INTERVIEWDATE condition
    const filteredRecords = records.filter((record: any) => {
      const interviewDate = new Date(record['INTERVIEWDATE']);
      return (
        record['L1 Interview Status'] === 'Selected for L2' &&
        record['COMPANYNAME'] === 'Capgemini India' &&
        interviewDate >= fiveDaysPrior
      );
    });

    // Select only the required fields that exist in the records
    const existingFields = REQUIRED_FIELDS.filter(field => 
      filteredRecords.length > 0 && field in filteredRecords[0]
    );

    // Create worksheet with filtered data
    const worksheet = XLSX.utils.json_to_sheet(
      filteredRecords.map((record: any) => {
        const filteredRecord: any = {};
        existingFields.forEach(field => {
          filteredRecord[field] = record[field];
        });
        return filteredRecord;
      })
    );

    // Create workbook and add worksheet
    const workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, 'L2 Selected Candidates');

    // Generate Excel file in memory
    const filename = 'l2Selected.xlsx';
    const excelBuffer = XLSX.write(workbook, { type: 'buffer', bookType: 'xlsx' });
    
    console.log('Excel file generated successfully in memory');
    
    // Create JSON data with filtered records in memory
    const jsonBuffer = Buffer.from(JSON.stringify(filteredRecords, null, 2));
    console.log('JSON data created in memory');
    
    return {
      success: true,
      filename,
      buffer: excelBuffer,
      record_count: filteredRecords.length,
      total_records: records.length,
      filtered_records: filteredRecords.length,
      fields_included: existingFields,
      missing_fields: REQUIRED_FIELDS.filter(field => !existingFields.includes(field))
    };
  } catch (error) {
    console.error('Error generating Excel:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error occurred'
    };
  }
} 